### PR TITLE
Updated students type of test

### PR DESCRIPTION
**Pull request by pralhad88**

### CODE REVIEW

File: lib/migrations/20190530163257_track_typeOfTest.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    await knex.schema.table('students', function(t){
+        t.string('typeOfTest').after('stage').nullable();
+    });
+
+    let students = await knex('students').select('*');
+    let promises = [];
+    
+    _.each(students, (student) => {
+        if (student.partnerId) {
+            promise = knex('students').update({
```
Comment (amarkrsinha1997) [inline lib/migrations/20190530163257_track_typeOfTest.js]: You need to specify the for which student are you updating it, or all the student would be updated every time.

File: lib/migrations/20190530163257_track_typeOfTest.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+                typeOfTest: "offlineTest"
+            })
+            promises.push(promise)
+        }else{
+            promise = knex('students').update({
```
Comment (amarkrsinha1997) [inline lib/migrations/20190530163257_track_typeOfTest.js]: Here also the same.

File: lib/models/student.js [hunk 3]
```diff
@@ -42,6 +42,9 @@ module.exports = class Student extends Model {
             percentageIn12th: Joi.string().allow(null),
             mathMarksIn12th: Joi.string().allow(null),
 
+            // Record students type of test
+            typeOfTest: Joi.string().allow(null),
```
Comment (amarkrsinha1997) [inline lib/models/student.js]: Don't Allow the typeOfTest to be null it should be either offline or online

Comment (amarkrsinha1997) [reply lib/models/student.js]: You can create an enum in constant.js

File: lib/services/assessment.js [hunk 4]
```diff
@@ -357,6 +357,10 @@ module.exports = class AssessmentService extends Schmervice.Service {
 
     async patchStudentDetails(key, details={}, txn=null) {
         const { Student } = this.server.models();
+
+        //update type of test
+        details.typeOfTest = "onlineTest";
```
Comment (amarkrsinha1997) [inline lib/services/assessment.js]: patchStudentDetails is also used in google sheet. So, if I change the student partner or student actual stage the function would be called.
So putting it here would be a good idea. Can we move it to somewhere else which is more related to online test specific?

File: lib/migrations/20190530163257_track_typeOfTest.js (hunk 2 above)
Comment (amarkrsinha1997) [reply lib/migrations/20190530163257_track_typeOfTest.js]: Also, do check the DB once to double confirm if the changes are working as expected.

File: lib/migrations/20190530163257_track_typeOfTest.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
+const _ = require('underscore');
+
+exports.up = async function(knex, Promise) {
+    await knex.schema.table('students', function(t){
+        t.string('typeOfTest').after('stage').nullable();
```
Comment (amarkrsinha1997) [inline lib/migrations/20190530163257_track_typeOfTest.js]: We can make it enum type either of online or offline

File: lib/migrations/20190530163257_track_typeOfTest.js [hunk 6]
```diff
@@ -0,0 +1,31 @@
+const _ = require('underscore');
+
+exports.up = async function(knex, Promise) {
+    await knex.schema.table('students', function(t){
```
Comment (amarkrsinha1997) [inline lib/migrations/20190530163257_track_typeOfTest.js]: does use knex.schema.table work? or do we need to use knex.schema.alterTable just double check it

File: lib/migrations/20190530163257_track_typeOfTest.js (hunk 1 above)
Comment (pralhad88) [reply lib/migrations/20190530163257_track_typeOfTest.js]: you means comment or anything else?

File: lib/migrations/20190530163257_track_typeOfTest.js (hunk 5 above)
Comment (pralhad88) [reply lib/migrations/20190530163257_track_typeOfTest.js]: how to make enum type in migration file

File: lib/migrations/20190530163257_track_typeOfTest.js (hunk 6 above)
Comment (pralhad88) [reply lib/migrations/20190530163257_track_typeOfTest.js]: knex.schema.alterTable it is work same as knex.schema.table is working.

File: lib/models/student.js (hunk 3 above)
Comment (pralhad88) [reply lib/models/student.js]: it is done.

File: lib/migrations/20190530163257_track_typeOfTest.js (hunk 2 above)
Comment (pralhad88) [reply lib/migrations/20190530163257_track_typeOfTest.js]: changes is working properly

File: lib/services/assessment.js (hunk 4 above)
Comment (pralhad88) [reply lib/services/assessment.js]: I am not getting what are you trying to say that

File: lib/migrations/20190530163257_track_typeOfTest.js (hunk 5 above)
Comment (amarkrsinha1997) [reply lib/migrations/20190530163257_track_typeOfTest.js]: I was telling in for constants.js my bad.

File: lib/migrations/20190530163257_track_typeOfTest.js (hunk 1 above)
Comment (amarkrsinha1997) [reply lib/migrations/20190530163257_track_typeOfTest.js]: studentId for which student are your updating it?

File: lib/migrations/20190530163257_track_typeOfTest.js (hunk 2 above)
Comment (amarkrsinha1997) [reply lib/migrations/20190530163257_track_typeOfTest.js]: I don't think so

Comment (pralhad88) [reply lib/migrations/20190530163257_track_typeOfTest.js]: now it's working properly.

File: lib/migrations/20190530163257_track_typeOfTest.js (hunk 1 above)
Comment (pralhad88) [reply lib/migrations/20190530163257_track_typeOfTest.js]: it is done

**Comment by RishabhVerma** [comment]
But will this logic work when the student stage is not `completedTest`

**Comment by RishabhVerma** [comment]
?

File: lib/services/partner.js [hunk 7]
```diff
@@ -27,6 +27,7 @@ internals.csvRowAttrsSchema = Joi.object({
     currentStatus: Joi.string().valid( ..._.keys(CONSTANTS.studentDetails.currentStatus) ).allow(null),
     caste: Joi.string().valid( ..._.keys(CONSTANTS.studentDetails.caste) ).allow(null),
     religon: Joi.string().valid( ..._.keys(CONSTANTS.studentDetails.religon) ).allow(null),
+    typeOfTest: Joi.string().allow(CONSTANTS.typeOfTest),
```
Comment (amarkrsinha1997) [inline lib/services/partner.js]: Make it default to offlineTest

Comment (amarkrsinha1997) [reply lib/services/partner.js]: @pralhad88

**Pull request merged**